### PR TITLE
Add type hints to synmark.

### PR DESCRIPTION
**Pull request by clokep**
Was trying to figure out synmark the other day and type hints were helpful.

### CODE REVIEW

File: synmark/__main__.py [hunk 1]
```diff
@@ -12,54 +12,60 @@
...
-    def _main(loops):
+    def _main(loops: int) -> float:
         reactor = make_reactor()
 
         file_out = StringIO()
         with redirect_stderr(file_out):
-            d = Deferred()
-            d.addCallback(lambda _: ensureDeferred(main(reactor, loops)))
+            d: "Deferred[float]" = Deferred()
+            d.addCallback(lambda _: ensureDeferred(main(reactor, loops)))  # type: ignore
```
Comment (DMRobertson) [inline synmark/__main__.py:48]: What's the ignore for here? Can we ignore a specific error code only?

File: synmark/__main__.py [hunk 2]
```diff
@@ -12,54 +12,60 @@
...
                 reactor.stop()
-                return _
+                return res
 
             d.addBoth(on_done)
             reactor.callWhenRunning(lambda: d.callback(True))
             reactor.run()
 
-        return d.result
+        return cast(float, d.result)
```
Comment (DMRobertson) [inline synmark/__main__.py:61]: Surprised a cast is needed here. Any idea why?

File: synmark/suites/logging.py [hunk 3]
```diff
@@ -11,60 +11,63 @@
...
 from synapse.config.logger import _setup_stdlib_logging
 from synapse.logging import RemoteHandler
+from synapse.synapse_rust import reset_logging_config
+from synapse.types import ISynapseReactor
 from synapse.util import Clock
 
 
 class LineCounter(LineOnlyReceiver):
     delimiter = b"\n"
+    count = 0
```
Comment (DMRobertson) [inline synmark/suites/logging.py:38]: Is this safe? `count` is now a class attribute. If something else modifies `self.count`, does it overwrite the class attribute or make a new instance attribute?

File: synmark/suites/logging.py [hunk 4]
```diff
@@ -11,60 +11,63 @@
...
+    logger_factory = Factory()
     logger_factory.wait_for = loops
     logger_factory.on_done = Deferred()
-    port = reactor.listenTCP(0, logger_factory, interface="127.0.0.1")
+    port = reactor.listenTCP(0, logger_factory, backlog=50, interface="127.0.0.1")
 
     # A fake homeserver config.
     class Config:
-        server_name = "synmark-" + str(loops)
-        no_redirect_stdio = True
```
Comment (DMRobertson) [inline synmark/suites/logging.py:67]: Is this a config option that we removed?

File: synmark/__main__.py (hunk 1 above)
Comment (clokep) [reply synmark/__main__.py:48]: Arg, it is `ensureDeferred` complaining about getting a coroutine, I think? Will double check.

File: synmark/suites/logging.py (hunk 3 above)
Comment (clokep) [reply synmark/suites/logging.py:38]: It isn't mutable so this is safe:

```pycon
>>> class A:
...   count = 0
... 
>>> a = A()
>>> b = A()
>>> c = A()
>>> a.count, b.count, c.count
(0, 0, 0)
>>> a.count = 3
>>> a.count, b.count, c.count
(3, 0, 0)
>>> c.count += 1
>>> a.count, b.count, c.count
(3, 0, 1)
```

File: synmark/suites/logging.py (hunk 4 above)
Comment (clokep) [reply synmark/suites/logging.py:67]: It is part of the logging config, I guess I should add it back there. 👍

File: synmark/__main__.py (hunk 2 above)
Comment (clokep) [reply synmark/__main__.py:61]:
```
synmark/__main__.py:61: error: Incompatible return value type (got "object", expected "float")  [return-value]
```

I think the `Deferred` isn't being properly marked as returning a `float`? (Maybe the Twisted trunk fixes this?)

Comment (clokep) [reply synmark/__main__.py]: I couldn't quite figure this out, I think twisted trunk would fix it though.

File: synmark/__main__.py [hunk 5]
```diff
@@ -12,54 +12,61 @@
...
+            def on_done(res: T) -> T:
+                if isinstance(res, Failure):
+                    res.printTraceback()
                     print(file_out.getvalue())
                 reactor.stop()
-                return _
+                return res
 
             d.addBoth(on_done)
             reactor.callWhenRunning(lambda: d.callback(True))
```
Comment (DMRobertson) [inline synmark/__main__.py:58]: Surprised mypy doesn't complain about feeding a bool into a Deferred[float]!

**Review by DMRobertson — APPROVED** [review]
I'm happy, provided you've tested that synmark still actually runs!

**Comment by clokep** [comment]
I have! The logging suite seems broken unrelated to this PR so I Don't want to hold this up.

**Pull request merged**